packages/astro/src/runtime/client/dev-overlay/overlay.ts

```diff
@@ -0,0 +1,457 @@
+	eventTarget: EventTarget;
+};
+
+document.addEventListener('DOMContentLoaded', async () => {
+	const WS_EVENT_NAME = 'astro-dev-overlay';
+
+	const builtinPlugins: DevOverlayPlugin[] = [
+		astroDevToolPlugin,
+		astroXrayPlugin,
+		astroAuditPlugin,
+	].map((plugin) => ({
+		...plugin,
+		active: false,
+		status: 'loading',
+		eventTarget: new EventTarget(),
+	}));
+
+	const customPluginsImports = (await loadDevOverlayPlugins()) as DevOverlayPluginDefinition[];
+	const customPlugins: DevOverlayPlugin[] = [];
+	customPlugins.push(
+		...customPluginsImports.map((plugin) => ({
+			...plugin,
+			active: false,
+			status: 'loading' as const,
+			eventTarget: new EventTarget(),
+		}))
+	);
+
+	const plugins: DevOverlayPlugin[] = [...builtinPlugins, ...customPlugins];
+
+	for (const plugin of plugins) {
+		plugin.eventTarget.addEventListener('plugin-notification', (evt) => {
+			const target = overlay.shadowRoot?.querySelector(`[data-plugin-id="${plugin.id}"]`);
+			if (!target) return;
+
+			let newState = true;
+			if (evt instanceof CustomEvent) {
+				newState = evt.detail.state ?? true;
+			}
+
+			target.querySelector('.notification')?.toggleAttribute('data-active', newState);
+		});
+	}
+
+	class AstroDevOverlay extends HTMLElement {
+		shadowRoot: ShadowRoot;
+		hoverTimeout: number | undefined;
+		isHidden: () => boolean = () => this.devOverlay?.hasAttribute('data-hidden') ?? true;
+		devOverlay: HTMLDivElement | undefined;
+
+		constructor() {
+			super();
+			this.shadowRoot = this.attachShadow({ mode: 'closed' });
+		}
+
+		// connect component
+		async connectedCallback() {
+			this.shadowRoot.innerHTML = `
+    <style>
+			#dev-overlay {
+				position: fixed;
+				bottom: 7.5%;
+				left: calc(50% + 32px);
+				transform: translate(-50%, 0%);
+				z-index: 999999;
+				display: flex;
+				gap: 8px;
+				align-items: center;
+				transition: bottom 0.2s ease-in-out;
+				pointer-events: none;
+			}
+
+			#dev-overlay[data-hidden] {
+				bottom: -40px;
+			}
+
+			#dev-overlay[data-hidden]:hover, #dev-overlay[data-hidden]:focus-within {
+				bottom: -35px;
+				cursor: pointer;
+			}
+
+			#dev-overlay[data-hidden] #minimize-button {
+				visibility: hidden;
+			}
+
+      #dev-bar {
+				height: 56px;
+				overflow: hidden;
+				pointer-events: auto;
+
+				background: linear-gradient(180deg, #13151A 0%, rgba(19, 21, 26, 0.88) 100%);
+				box-shadow: 0px 0px 0px 0px #13151A4D;
+				border: 1px solid #343841;
+				border-radius: 9999px;
+			}
+
+			#dev-bar .item {
+				display: flex;
+				justify-content: center;
+				align-items: center;
+				width: 64px;
+				border: 0;
+				background: transparent;
+				color: white;
+				font-family: system-ui, sans-serif;
+				font-size: 1rem;
+				line-height: 1.2;
+				white-space: nowrap;
+				text-decoration: none;
+				padding: 0;
+				margin: 0;
+			}
+
+			#dev-bar #bar-container .item:hover, #dev-bar #bar-container .item:focus {
+				background: rgba(27, 30, 36, 1);
+				cursor: pointer;
```

Comment:
Nit, but focus outline is not always visible on the overlay buttons because it’s cut off by the `overflow: hidden` on `#dev-bar`.

For example, here the first item is focused, but the outline is cut off and then hidden behind the selected second item, the subtle background-color change is too small to clearly convey focus:

<img width="215" alt="Astro Dev Overlay" src="https://github.com/withastro/astro/assets/357379/42e03cdb-e6a2-49e5-b897-a705d7ad5087">

We could inset the outline to ensure it’s visible:

```suggestion
				cursor: pointer;
				outline-offset: -3px;
```

Reply:
Oh weird, on Chrome the outline is always visible (at least, on my computer)